### PR TITLE
#25 feat(contract): implement card energy logic 

### LINKED ISSUE #25: Implement card energy logic
# Description

As per the game rules each card costs a certain amount of energy. The energy is reset and incremented at each turn. Each player starts with 1 energy and is incremented each turn.

# Acceptance criteria

* Each player starts with 1 energy
* Energy is incremented by 1 at each of your turns
* Energy is reset at each of your turns

### PR DESCRIPTION
**Pull request by TeddyNotBear**
Reset the energy of player at the end of each turn.
Fix end_turn test, the game id was incorrect, it should not be equal to zero but to pedersen(player1, player2)
Fixes #25

### CODE REVIEW

File: contracts/src/systems/end_turn.cairo [hunk 1]
```diff
@@ -2,14 +2,13 @@
...
-    use tsubasa::components::Game;
+    use tsubasa::components::{Game, Energy};
     use tsubasa::events::{emit, EndTurn};
     use array::ArrayTrait;
 
 
     fn execute(ctx: Context, game_id: felt252) {
-        let game = get!(ctx.world, game_id, Game);
-
+        let mut game = get!(ctx.world, game_id, Game);
```
Comment (LucasLvy) [inline contracts/src/systems/end_turn.cairo:11]:
```suggestion
        let game = get!(ctx.world, game_id, Game);
```

File: contracts/src/systems/end_turn.cairo [hunk 2]
```diff
@@ -21,6 +20,8 @@ mod end_turn_system {
                 outcome: game.outcome
             }
         );
+        game = get!(ctx.world, game_id, Game);
```
Comment (LucasLvy) [inline contracts/src/systems/end_turn.cairo:23]:
```suggestion
        let game = get!(ctx.world, game_id, Game);
```

File: contracts/src/systems/end_turn.cairo [hunk 3]
```diff
@@ -21,6 +20,8 @@ mod end_turn_system {
...
+        set!(ctx.world, Energy { game_id: game_id, player: ctx.origin, remaining: game.turn + 1 });
```
Comment (LucasLvy) [inline contracts/src/systems/end_turn.cairo:24]: hmm no turn 2 both player should have 1 energy, turn 4 both should have 2
```suggestion
        set!(ctx.world, Energy { game_id: game_id, player: ctx.origin, remaining: game.turn / 2_u128 });
```
Make sure that you change the turns into u128 to have a proper division

**Review by LucasLvy — CHANGES_REQUESTED** [review]

**Comment by LucasLvy** [comment]
format also plz

File: contracts/src/systems/end_turn.cairo (hunk 3 above)
Comment (TeddyNotBear) [reply contracts/src/systems/end_turn.cairo:24]: Can you explain to me how this should work, maybe i misunderstood the process of turns. I thought it was:

Turn 0 =>
- Player 1
    - Start with 1 energy
    - End of turn increment remaining energy by 1. Energy.remaining should not be equal to 2.
- Player 2
    - Start with 1 energy
    - End of turn increment remaining energy by 1. Energy.remaining should not be equal to 2.

Turn 1 =>
- Player 1
    - Start new turn with 2 energy
    - End of turn increment by remaining energy 1. Energy.remaining should not be equal to 3.
- Player 2
    - Start new turn with 2 energy
    - End of turn increment by remaining energy 1. Energy.remaining should not be equal to 3.

Comment (LucasLvy) [reply contracts/src/systems/end_turn.cairo:24]: ooooh i see what's happening, each time a player presses end_turn, the turn number is incremented. So at turn 8 for example each player will have played 4 turns and should have 4 energy. Does that solve the confusion ?

**Comment by LucasLvy** [comment]
run scarb fmt plz

File: contracts/src/tests/end_turn.cairo [hunk 4]
```diff
@@ -56,9 +56,15 @@ fn test_end_turn() {
...
+    assert(game.game_id == expected_game.game_id, 'invalid game_id');
     assert(game.player1_score == expected_game.player1_score, 'Wrong player1 score');
     assert(game.player2_score == expected_game.player2_score, 'Wrong player2 score');
     assert(game.turn == expected_game.turn, 'Wrong turn value');
     // Check that option is None
     assert(game.outcome.is_none(), 'Wrong outcome value');
+
+    let expected_energy = Energy { game_id: game_id, player: player1, remaining: 1 };
+    let player_energy = get!(world, (expected_energy.game_id, expected_energy.player), Energy);
+    //Check that player energy is correclty incremented at the end of each turn.
```
Comment (LucasLvy) [inline contracts/src/tests/end_turn.cairo:68]:
```suggestion
    // Check that player energy is correclty incremented at the end of each turn.
```

File: contracts/src/tests/end_turn.cairo [hunk 5]
```diff
@@ -56,9 +56,15 @@ fn test_end_turn() {
...
     };
 
+    assert(game.game_id == expected_game.game_id, 'invalid game_id');
     assert(game.player1_score == expected_game.player1_score, 'Wrong player1 score');
     assert(game.player2_score == expected_game.player2_score, 'Wrong player2 score');
     assert(game.turn == expected_game.turn, 'Wrong turn value');
     // Check that option is None
     assert(game.outcome.is_none(), 'Wrong outcome value');
+
+    let expected_energy = Energy { game_id: game_id, player: player1, remaining: 1 };
```
Comment (LucasLvy) [inline contracts/src/tests/end_turn.cairo:66]: Player 1 should have 2 energy after the end of turn 0
```suggestion
    let expected_energy = Energy { game_id: game_id, player: player1, remaining: 2 };
```

**Review by LucasLvy — CHANGES_REQUESTED** [review]

**Review by LucasLvy — APPROVED** [review]
lgtm

**Pull request merged**